In [130]:
import pickle
from catboost import CatBoostClassifier, Pool, metrics, cv
import numpy as np
import pandas as pd
from category_encoders.cat_boost import CatBoostEncoder
from sklearn.preprocessing import OrdinalEncoder
from sklearn.metrics import cohen_kappa_score, f1_score, matthews_corrcoef, recall_score, precision_score
# from category_encoders.m_estimate import MEstimateEncoder
orders = catalog.load("orders_features")
parameters = catalog.load("parameters")
orders_full = catalog.load("orders_full")
parameters = catalog.load("parameters")
enc_frame = catalog.load("enc_frame")
enc_cols = catalog.load("enc_cols")

2022-03-18 16:20:18,050 - kedro.io.data_catalog - INFO - Loading data from `orders_features` (PickleDataSet)...
2022-03-18 16:20:18,750 - kedro.io.data_catalog - INFO - Loading data from `parameters` (MemoryDataSet)...
2022-03-18 16:20:18,751 - kedro.io.data_catalog - INFO - Loading data from `orders_full` (PickleDataSet)...
2022-03-18 16:20:18,941 - kedro.io.data_catalog - INFO - Loading data from `parameters` (MemoryDataSet)...
2022-03-18 16:20:18,942 - kedro.io.data_catalog - INFO - Loading data from `enc_frame` (PickleDataSet)...
2022-03-18 16:20:19,270 - kedro.io.data_catalog - INFO - Loading data from `enc_cols` (PickleDataSet)...


In [178]:
budget_incs = 

train

weekdayOrdered weekdayReceived weDelivery accOldSys  color  size  \
0                   6               1          0         0     40   108   
1                   6               1          0         0     66   102   
2                   6               1          0         0     33   102   
3                   0               3          0         0     47   108   
4                   0               3          0         1     17    59   
...               ...             ...        ...       ...    ...   ...   
480944              6               1          0         0     18    60   
480945              6               1          0         0     48    59   
480946              6               2          0         0     17   113   
480947              6               2          0         0     17   113   
480948              6               2          0         0     76   113   

       userAgeImp  customerID manufacturerID  salutation  ...  \
0               0         422             25           3  ...   
1               0         422             21           3  ...   
2               0         422             21           3  ...   
3               0         430             14           3  ...   
4               0         445             53           3  ...   
...           ...         ...            ...         ...  ...   
480944          0        4657              1           2  ...   
480945          1       59731             18           3  ...   
480946          0       59732              5           2  ...   
480947          0       59732             49           2  ...   
480948          0       59732             49           2  ...   

        100_100_False_lift_base  100_100_False_conf_base      10_m      20_m  \
0                      0.632280                 0.632280  0.638086  0.633387   
1                      0.664222                 0.664222  0.608780  0.607564   
2                      0.664222                 0.664222  0.608780  0.607564   
3                      0.447011                 0.447011  0.447171  0.447330   
4                      0.536452                 0.536452  0.570911  0.558466   
...                         ...                      ...       ...       ...   
480944                 0.661758                 0.661758  0.647952  0.636466   
480945                 0.616216                 0.616216  0.609362  0.603176   
480946                 0.617907                 0.617907  0.620669  0.607880   
480947                 0.690917                 0.690917  0.675380  0.661896   
480948                 0.739875                 0.739875  0.735832  0.727007   

            50_m     100_m  10_m_smooth  20_m_smooth  50_m_smooth  \
0       0.620852  0.604028     0.653933     0.649118     0.636272   
1       0.604052  0.598618     0.604571     0.603363     0.599876   
2       0.604052  0.598618     0.628495     0.627239     0.623614   
3       0.447798  0.448551     0.481224     0.481395     0.481898   
4       0.535918  0.518248     0.564227     0.551928     0.529644   
...          ...       ...          ...          ...          ...   
480944  0.609929  0.581497     0.675464     0.663491     0.635827   
480945  0.587778  0.569317     0.596218     0.590166     0.575099   
480946  0.580636  0.554550     0.600010     0.587647     0.561309   
480947  0.630796  0.597558     0.687281     0.673559     0.641911   
480948  0.703873  0.673721     0.708995     0.700491     0.678201   

        100_m_smooth  
0           0.619029  
1           0.594480  
2           0.618004  
3           0.482709  
4           0.512180  
...              ...  
480944      0.606188  
480945      0.557037  
480946      0.536092  
480947      0.608087  
480948      0.649148  

[480949 rows x 177 columns]

In [183]:
budget_dict = {}
budget_dict["train"] = {}
budget_dict["test"] = {}

for budget in data["budgets"].unique():
    budget_dict["train"][budget] = train[train["budgets"]<=budget].index
    budget_dict["test"][budget] = test[test["budgets"]<=budget].index
self.budget_dict = budget_dict

NameError: name 'self' is not defined

In [186]:
train.loc[budget_dict["train"][8]]

weekdayOrdered weekdayReceived weDelivery accOldSys  color  size  \
0                   6               1          0         0     40   108   
1                   6               1          0         0     66   102   
2                   6               1          0         0     33   102   
5                   0               3          0         1     21   113   
7                   0               1          0         1     21   113   
...               ...             ...        ...       ...    ...   ...   
480941              6               3          0         0      7   113   
480942              6               2          0         0     80   113   
480945              6               1          0         0     48    59   
480946              6               2          0         0     17   113   
480947              6               2          0         0     17   113   

       userAgeImp  customerID manufacturerID  salutation  ...  \
0               0         422             25           3  ...   
1               0         422             21           3  ...   
2               0         422             21           3  ...   
5               0         445             87           3  ...   
7               0         464              3           3  ...   
...           ...         ...            ...         ...  ...   
480941          0       59730              3           3  ...   
480942          0       59730              3           3  ...   
480945          1       59731             18           3  ...   
480946          0       59732              5           2  ...   
480947          0       59732             49           2  ...   

        100_100_False_lift_base  100_100_False_conf_base      10_m      20_m  \
0                      0.632280                 0.632280  0.638086  0.633387   
1                      0.664222                 0.664222  0.608780  0.607564   
2                      0.664222                 0.664222  0.608780  0.607564   
5                      0.765985                 0.765985  0.748821  0.731969   
7                      0.392427                 0.392427  0.352728  0.353090   
...                         ...                      ...       ...       ...   
480941                 0.343625                 0.343625  0.333553  0.333949   
480942                 0.343625                 0.343625  0.333553  0.333949   
480945                 0.616216                 0.616216  0.609362  0.603176   
480946                 0.617907                 0.617907  0.620669  0.607880   
480947                 0.690917                 0.690917  0.675380  0.661896   

            50_m     100_m  10_m_smooth  20_m_smooth  50_m_smooth  \
0       0.620852  0.604028     0.653933     0.649118     0.636272   
1       0.604052  0.598618     0.604571     0.603363     0.599876   
2       0.604052  0.598618     0.628495     0.627239     0.623614   
5       0.692169  0.648132     0.740055     0.723400     0.684066   
7       0.354162  0.355909     0.366263     0.366638     0.367751   
...          ...       ...          ...          ...          ...   
480941  0.335125  0.337043     0.303533     0.303893     0.304962   
480942  0.335125  0.337043     0.352808     0.353227     0.354470   
480945  0.587778  0.569317     0.596218     0.590166     0.575099   
480946  0.580636  0.554550     0.600010     0.587647     0.561309   
480947  0.630796  0.597558     0.687281     0.673559     0.641911   

        100_m_smooth  
0           0.619029  
1           0.594480  
2           0.618004  
5           0.640544  
7           0.369566  
...              ...  
480941      0.306708  
480942      0.356499  
480945      0.557037  
480946      0.536092  
480947      0.608087  

[387820 rows x 177 columns]

In [181]:
incdict = {}
for budget in data["budgets"].unique():
    incdict[budget] = data[data["budgets"]<=budget].index

In [140]:
data = orders["train_test"]
ord_encode =  parameters["coldict"]["ord_encode"]
# data[ord_encode]
train = data[data["val_set"]==0]
test = data[data["val_set"]==1]
enc = OrdinalEncoder(handle_unknown= "use_encoded_value", unknown_value=np.nan) # )
enc_train = enc.fit_transform(train[ord_encode])
enc_test = enc.transform(test[ord_encode])
data.loc[:, ord_encode] = np.nan
data.loc[train.index, ord_encode] = enc_train
data.loc[test.index, ord_encode] = enc_test
for column in ord_encode:
    data[column].fillna(max(data[column])-1, inplace=True)
    data[column] = data[column].astype(int)

In [132]:
from torch.utils.data import Dataset, DataLoader #, TensorDataset
import torch
import torch.nn.functional as F
import torch.nn as nn
class TabularDataset(Dataset):
    def __init__(self, X, y, cat_feat=[], cont_feat=[]):
        """
        # https://yashuseth.blog/2018/07/22/pytorch-neural-network-for-tabular-data-with-categorical-embeddings/
        Characterizes a Dataset for PyTorch
        Parameters
        ----------
        X: array with features
        y: 1D array with explainable variable
        cat_feat: int or List of ints with incides
        The names of the categorical columns in the data.
        These columns will be passed through the embedding
        layers in the model. These columns must be
        label encoded beforehand. 
        output_col: string
        The name of the output variable column in the data
        provided.
        """
        self.n = X.shape[0]

        self.y = torch.from_numpy(np.array(y.astype(np.float32))).reshape(-1, 1)

        self.cat_feat = cat_feat # if any(cat_feat) else []
        self.cont_feat = cont_feat # if cont_feat else []

        self.cont_X = torch.from_numpy(
            X[:, cont_feat].astype(np.float32) if any(self.cont_feat) else np.zeros((self.n, 1)))
        self.cat_X = torch.from_numpy(
            X[:, cat_feat].astype(np.int32) if any(self.cat_feat) else np.zeros((self.n, 1)))

    def __len__(self):
        """
        Denotes the total number of samples.
        """
        return self.n

    def __getitem__(self, idx):
        """
        Generates one sample of data.
        """
        return [self.y[idx], self.cont_X[idx], self.cat_X[idx]]

In [133]:
cols_dict = parameters["coldict"]
cat_cols = cols_dict['cat_pass'] + cols_dict['cat_catboost']
cat_dims = [len(np.unique(X_train[:, inc])) for inc in range(len(cat_cols))] # 
emb_dims = [(x + 1, min(600, round(1.6 * x **0.56))) for x in cat_dims]
embeds = nn.ModuleList([nn.Embedding(x, y)
                                         for x, y in emb_dims])

In [134]:
train = data[data["val_set"]==0]
test = data[data["val_set"]==1]

X_train = train[cat_cols + cols_dict["cont"]].to_numpy()
X_test = test[cat_cols + cols_dict["cont"]].to_numpy()

y_train = train["returnShipment"].to_numpy()
y_test = test["returnShipment"].to_numpy()

In [135]:
# train_dataset = TabularDataset(X_train, y_train, cat_inc, cont_inc)
cat_inc = np.array(range(len(cat_cols)))
cont_inc = np.array(range(len(cat_cols), X_train.shape[1]))

train_dataset = TabularDataset(X_train, y_train, cat_inc, cont_inc)
train_loader = torch.utils.data.DataLoader(dataset=train_dataset, batch_size=1024)
# test_loader = torch.utils.data.DataLoader(dataset=test_dataset, batch_size=1024)

In [149]:
data.loc[:, cat_cols] = data[cat_cols].apply(lambda x: pd.to_numeric(x))

In [158]:
ord_encode= ["color", "size", "customerID", "salutation", "state", "item_color_cat", "item_size_cat", "itemID", "manufacturerID"]
# cat_passnoenc: ['weekdayOrdered', 'weekdayReceived', 'weDelivery', 'accOldSys', 'userAgeImp']
cat_pass= ['weekdayOrdered', 'weekdayReceived', 'weDelivery', 'accOldSys', "color", "size", "userAgeImp"]
# double: one normal + one catboost col; seperate for arguments -> ["manufacturerID", "salutation", "state"] + ['item_color_cat', 'item_size_cat']
cat_catboost= ["customerID", "manufacturerID", "salutation", "state", 'item_color_cat', 'item_size_cat', 'monthOrdered', "itemID"] #

In [170]:
"""catcols = pd.Series(cat_pass+cat_catboost)
numeric = catcols[~catcols.isin(ord_encode)]
data.loc[:, numeric]
for i, (y, cont_x, cat_x) in enumerate(train_loader):
    print(y.shape, cont_x.shape, cat_x.shape)"""

'catcols = pd.Series(cat_pass+cat_catboost)\nnumeric = catcols[~catcols.isin(ord_encode)]\ndata.loc[:, numeric]'

In [126]:
vals = {}
for cat in cat_inc:
    vals[cat] = []
for i, (y, cont_x, cat_x) in enumerate(train_loader):
    for j in range(cat_x.shape[1]):
        vals[j] = np.unique(np.append(np.array(vals[j]), cat_x[:, j].numpy()))

In [ ]:
for i in range(15):
    print(np.max(np.array(cat_x[:, i])))

In [ ]:
# np.max(np.array(cat_x[:, i]))

In [171]:
emb_dims = [(8, 5), (7, 4), (3, 2), (3, 2), (86, 19), (116, 23), (3, 2), (54834, 600), (164, 28), (6, 4), (17, 8), (8187, 249), (15189, 351), (13, 6)]
# [emb_layer(cat_data[:, i]) for i, emb_layer in enumerate(embeds)]

In [12]:
for column in ord_encode:
    print(column, data[column].nunique(), data[column].min(), data[column].max())

color 85 0 84
size 115 0 114
customerID 59733 0 59732
salutation 5 0 4
state 16 0 15
item_color_cat 8230 0 8229
item_size_cat 15432 0 15431
itemID 3002 0 3001


In [60]:
from sklearn.preprocessing import OrdinalEncoder
for column in data.columns:
    
enc = OrdinalEncoder(handle_unknown= "use_encoded_value", unknown_value=np.nan) # )
enc_ord = enc.fit_transform(data[ord_encode])
data.loc[:, label_encode] = enc_ord

In [ ]:
label_encode: ["color", "size", "customerID", "salutation", "state", "item_color_cat", "item_size_cat", "itemID"]
cat_pass: ['weekdayOrdered', 'weekdayReceived', 'weDelivery', 'accOldSys', 'userAgeImp']

In [3]:
cols_dict = parameters["coldict"]
cat_cols = cols_dict['cat_pass'] + cols_dict['cat_catboost']
cat_cols.remove(parameters["artnr_col"]) # remove art and add it depending on config
cont_cols = cols_dict["cont"]
# orders["train_val"][cat_cols]

weekdayOrdered weekdayReceived weDelivery accOldSys  color size  \
0                   6               1          0         0  denim    m   
1                   6               1          0         0  ocher   9+   
2                   6               1          0         0  curry   9+   
3                   0               3          0         0  green    m   
4                   0               3          0         1  black   39   
...               ...             ...        ...       ...    ...  ...   
480944              6               1          0         0   blue  39+   
480945              6               1          0         0   grey   39   
480946              6               2          0         0  black  xxl   
480947              6               2          0         0  black  xxl   
480948              6               2          0         0    red  xxl   

       userAgeImp customerID manufacturerID salutation                 state  \
0               0        794             25        Mrs    Baden-Wuerttemberg   
1               0        794             21        Mrs    Baden-Wuerttemberg   
2               0        794             21        Mrs    Baden-Wuerttemberg   
3               0        808             14        Mrs                Saxony   
4               0        825             53        Mrs  Rhineland-Palatinate   
...           ...        ...            ...        ...                   ...   
480944          0       7032              1         Mr          Lower Saxony   
480945          1      86610             18        Mrs                Berlin   
480946          0      86611              5         Mr                Berlin   
480947          0      86611             49         Mr                Berlin   
480948          0      86611             49         Mr                Berlin   

       item_color_cat item_size_cat monthOrdered  
0           186_denim         186_m            4  
1            71_ocher         71_9+            4  
2            71_curry         71_9+            4  
3            22_green          22_m            4  
4           151_black        151_39            4  
...               ...           ...          ...  
480944      2561_blue      2561_39+            3  
480945      2669_grey       2669_39            3  
480946     2458_black      2458_xxl            3  
480947     2659_black      2659_xxl            3  
480948       2375_red      2375_xxl            3  

[480949 rows x 14 columns]

In [3]:
orders["train_test"].columns[:25]
# orders_full[orders_full['customerID']==878].returnShipment.mean()

Index(['weekdayOrdered', 'weekdayReceived', 'weDelivery', 'accOldSys', 'color',
       'size', 'userAgeImp', 'customerID', 'manufacturerID', 'salutation',
       'state', 'item_color_cat', 'item_size_cat', 'monthOrdered', 'itemID',
       'price', 'discount', 'basket_price', 'basketNArts', 'durationDelivery',
       'accAgeAtOrder', 'UserAgeOrder', 'returnShipment', 'val_set',
       'budgets'],
      dtype='object')

In [66]:
"""miniframe = orders["train_test"][pd.DataFrame(rest_features).columns.to_list()]
for col in miniframe.columns:
    print(miniframe[col].dtype)
for col in examp.columns:
    print(pd.DataFrame(examp)[col].dtype)"""

float64
category
category
float64
int64


In [68]:
examp = pd.DataFrame(rest_features)

In [73]:
for column in ['item_color_cat', 'item_size_cat']:
    examp[column] = examp[column].astype("category")

In [29]:
orders["train_test"][enc_frame.columns].equals(enc_frame)

True

### Check Data + Training
* time+other features equal to old version (old data still available)

In [15]:
data_val = orders["train_val"]
train_val = data_val[data_val["val_set"]==0]
test_val = data_val[data_val["val_set"]==1]
pos_val_train = train_val[train_val["returnShipment"]==1].index
neg_val_train = train_val[train_val["returnShipment"]==0].index
pos_val_test = test_val[test_val["returnShipment"]==1].index
neg_val_test = test_val[test_val["returnShipment"]==0].index

In [40]:
print(test_val["returnShipment"].mean(),
train_val["returnShipment"].mean())

0.5046327728743161

In [16]:
data_test = orders["train_test"]
train_val = data_val[data_val["val_set"]==0]
test_val = data_val[data_val["val_set"]==1]
pos_val_train = train_val[train_val["returnShipment"]==1].index
neg_val_train = train_val[train_val["returnShipment"]==0].index
pos_val_test = test_val[test_val["returnShipment"]==1].index
neg_val_test = test_val[test_val["returnShipment"]==0].index

In [51]:
# data_test["returnShipment"]
train_val.loc[neg_val_train, "100_1_False_conf"].mean(),
train_val.loc[pos_val_train, "100_1_False_conf"].mean(),
test_val.loc[pos_val_test, "100_1_False_conf"].mean(),
test_val.loc[neg_val_test, "100_1_False_conf"].mean()

0.4602073960213485

In [10]:
# print(test_val["returnShipment"].mean(),
# train_val["returnShipment"].mean())

In [4]:
depth = 8
l2_leaf_reg = 12
lr = 0.2620077982153718

retp_cols = ["100_1_True_conf"]

data_test = orders["train_test"]

cols_dict = parameters["coldict"]
cat_cols = cols_dict['cat_pass'] + cols_dict['cat_catboost']
cat_cols.remove(parameters["artnr_col"])
cont_cols = cols_dict["cont"]
output_col = parameters["target_col"]
cat_features = np.array(range(len(cat_cols)))
train_inc = data_test[data_test["val_set"]==0].index
test_inc = data_test[data_test["val_set"]==1].index

data_test = data_test[cat_cols+cont_cols+retp_cols+[output_col]]
X_train= data_test.drop(output_col, axis=1).loc[train_inc]
y_train = data_test.loc[:, output_col].loc[train_inc]
X_test= data_test.drop(output_col, axis=1).loc[test_inc]
y_test = data_test.loc[test_inc, output_col]

train_pool = Pool(X_train, y_train, cat_features=cat_features)
validate_pool = Pool(X_test, y_test, cat_features=cat_features)

In [6]:
data_test.columns[:20]

Index(['weekdayOrdered', 'weekdayReceived', 'weDelivery', 'accOldSys', 'color',
       'size', 'userAgeImp', 'customerID', 'manufacturerID', 'salutation',
       'state', 'item_color_cat', 'item_size_cat', 'monthOrdered', 'price',
       'discount', 'basket_price', 'basketNArts', 'durationDelivery',
       'accAgeAtOrder'],
      dtype='object')

In [8]:
params = {
    'iterations': 1000,
    'learning_rate': lr,
    'eval_metric': metrics.MCC(),
    'random_seed': 42,
    'logging_level': 'Silent',
    'use_best_model': True,
    'od_type': 'Iter',  # https://catboost.ai/en/docs/features/overfitting-detector-desc#od_wait
    'od_wait': 50,
    'allow_writing_files': False,
    'one_hot_max_size': 16, #
    'l2_leaf_reg': l2_leaf_reg,
    'depth': depth}
model = CatBoostClassifier(**params)
model.fit(train_pool, eval_set=validate_pool)

In [9]:
y_pred = model.predict(X_test)
MCC_test= matthews_corrcoef(y_test, y_pred)
MCC_test

0.34735266400831144

## Content
* plots: n_Warenkorb vs. ret%
* ret_p vs. ret% in test+train (feature check)

### Data Check (ret_p values etc.)
* check imputed values and ret_p>1
* dual vs. multi-item baskets
* generate_ret_p review
    * lookup dict creation (multi + art_df)
    * Lookup feature generation: get_ret_perc_train lookup
    
### Catboost Training

### Load data

In [4]:
orders = catalog.load("orders_full")
orders["order_month"] = pd.to_datetime(orders_full["orderDate"]).dt.month
orders["order_year"] = pd.to_datetime(orders_full["orderDate"]).dt.year
orders_features = catalog.load("orders_features")
orders["basketID"] = orders["customerID"].astype(str) + orders["orderDate"].astype(str)
train = orders[orders["val_set"]==0]
test = orders[orders["val_set"]==1]
# orders_features["basketID"] = orders_features["customerID"].astype(str) + orders_features["orderDate"].astype(str)
parameters = catalog.load("parameters")
#train = orders_features[orders_features["val_set"]==0]
#test = orders_features[orders_features["val_set"]==1]
# catboosted_cols = pd.Series(cols["cat_catboost"])+"Catboost"
# orders_features.loc[:, catboosted_cols] = orders_features[catboosted_cols]
output_col = parameters["coldict"]["output_col"]
coldict = parameters['coldict']
coldict['cat_catboost']

2022-02-12 12:58:56,147 - kedro.io.data_catalog - INFO - Loading data from `orders_full` (PickleDataSet)...
2022-02-12 12:58:56,476 - kedro.io.data_catalog - INFO - Loading data from `orders_features` (PickleDataSet)...
2022-02-12 12:58:56,908 - kedro.io.data_catalog - INFO - Loading data from `parameters` (MemoryDataSet)...


['manufacturerID',
 'salutation',
 'state',
 'item_color_cat',
 'item_size_cat',
 'monthOrdered',
 'itemID']

**plots for no_arts in comparison to return shipments**

In [131]:
# orders_features.groupby("basketNArts")["returnShipment"].mean().sort_index().plot()
# orders_features.groupby("basketNArts")["returnShipment"].mean().sort_index()
# orders_features.groupby("basketNArts")["returnShipment"].size().sort_index().cumsum()/orders_features.shape[0]

### check ret_ps from data: test vs. train
1. single-article vs. multi-article orders
2. ret_p imputed vs. non-imputed orders

In [5]:
# single-basket
print("train",
orders_features[(orders_features["basketNArts"]==1)&
               (orders_features["val_set"]==0)][output_col].mean(),
orders_features[(orders_features["basketNArts"]>1)&
               (orders_features["val_set"]==0)][output_col].mean(),
      "\n test",
orders_features[(orders_features["basketNArts"]==1)&
               (orders_features["val_set"]==1)][output_col].mean(),
orders_features[(orders_features["basketNArts"]>1)&
               (orders_features["val_set"]==1)][output_col].mean()
)
# fast 2% mehr Unterschied in Multi-basket

train 0.3255756497677636 0.4969793835263455 
 test 0.3317452357071214 0.5143415057496203


In [6]:
# imputed mean ret_p vs. non-imputed
print("train",
orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==1)&
               (orders_features["val_set"]==0)][output_col].mean(),
orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]!=1)&
               (orders_features["val_set"]==0)][output_col].mean(),
      "\n test",
orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==1)&
               (orders_features["val_set"]==1)][output_col].mean(),
orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]!=1)&
               (orders_features["val_set"]==1)][output_col].mean()
)

train 0.410765615198075 0.5030469388955113 
 test 0.43738247461451674 0.5166320166320166


#### Non-imputed ret_p sorted by lift and conf

In [7]:
i = 0    
print("train_no_rets",
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==0) &
               (orders_features[output_col]==i)]['min_5_top_5_sqrt_w_lift'].mean(),    
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==0) &
               (orders_features[output_col]==i+1)]['min_5_top_5_sqrt_w_lift'].mean(),
"\n test_no_rets",
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==1) &
               (orders_features[output_col]==i)]['min_5_top_5_sqrt_w_lift'].mean(),
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==1) &
               (orders_features[output_col]==i+1)]['min_5_top_5_sqrt_w_lift'].mean())
# no rets diff: 2,8; rets diff: .8%
# no_ret -> ret

train_no_rets 0.4679824144287679 0.5391884884846062 
 test_no_rets 0.49462302668362834 0.5472992986046764


In [35]:
i = 0    
print("train_no_rets",
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==0)][output_col].mean(),    
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==0)][output_col].mean(),
"\n test_no_rets",
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==1)][output_col].mean(),
orders_features[(orders_features["min_5_top_5_sqrt_w_lift_Imp"]==0)&
               (orders_features["val_set"]==1)][output_col].mean())
# no rets diff: 2,8; rets diff: .8%
# no_ret -> ret

train_no_rets 0.5030469388955113 0.5030469388955113 
 test_no_rets 0.5166320166320166 0.5166320166320166


In [9]:
i = 0    
print("train_no_rets",
orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==0)&
               (orders_features["val_set"]==0) &
               (orders_features[output_col]==i)][output_col].mean(),
    
orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==0)&
               (orders_features["val_set"]==0) &
               (orders_features[output_col]==i+1)]['min_5_top_5_sqrt_w_conf'].mean(),
"\n test_no_rets",
orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==0)&
               (orders_features["val_set"]==1) &
               (orders_features[output_col]==i)]['min_5_top_5_sqrt_w_conf'].mean(),

orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==0)&
               (orders_features["val_set"]==1) &
               (orders_features[output_col]==i+1)]['min_5_top_5_sqrt_w_conf'].mean())

train_no_rets 0.4669398372259073 0.5382890366105384 
 test_no_rets 0.49416179974181296 0.5467981679456653


### Data check
* check for imputed values
* check why some ret_p values are >5

In [166]:
arts_ret_p = catalog.load("arts_ret_p")
lookup_ret_perc_art = catalog.load("lookup_ret_perc_art")
lookup_basket_ret_p = catalog.load("lookup_basket_ret_p")
lookup_basket_occ = catalog.load('lookup_basket_occ')
lookup_lift_arts = catalog.load("lookup_lift_arts")
lookup_tot_ord = catalog.load("lookup_tot_ord")
arts_used_ret_p = catalog.load("arts_used_ret_p")
lookup_droped_ret_arts = catalog.load("lookup_droped_ret_arts")

2022-02-11 14:56:39,815 - kedro.io.data_catalog - INFO - Loading data from `arts_ret_p` (PickleDataSet)...
2022-02-11 14:56:39,824 - kedro.io.data_catalog - INFO - Loading data from `lookup_ret_perc_art` (PickleDataSet)...
2022-02-11 14:56:39,825 - kedro.io.data_catalog - INFO - Loading data from `lookup_basket_ret_p` (PickleDataSet)...
2022-02-11 14:56:39,922 - kedro.io.data_catalog - INFO - Loading data from `lookup_basket_occ` (PickleDataSet)...
2022-02-11 14:56:40,366 - kedro.io.data_catalog - INFO - Loading data from `lookup_lift_arts` (PickleDataSet)...
2022-02-11 14:56:40,475 - kedro.io.data_catalog - INFO - Loading data from `lookup_tot_ord` (PickleDataSet)...
2022-02-11 14:56:40,476 - kedro.io.data_catalog - INFO - Loading data from `arts_used_ret_p` (PickleDataSet)...
2022-02-11 14:56:40,484 - kedro.io.data_catalog - INFO - Loading data from `lookup_droped_ret_arts` (PickleDataSet)...


### imputed
* solved problem: own article got shown wrongly

In [174]:
len(lookup_droped_ret_arts[5])
# grouped = orders.groupby("itemID")["returnShipment"].mean()
# n_rets = grouped[grouped==0].index
#orders_features[(orders_features["itemID"].isin(lookup_droped_ret_arts[5]))
#               & (orders_features["basketNArts"]>1)]

In [175]:
# orders_features[(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==1)& (orders_features["itemID"].isin(lookup_droped_ret_arts[5]))]

In [164]:
# orders[orders["itemID"].isin(n_rets)]
(orders_features["min_5_top_5_sqrt_w_conf_Imp"]==1).sum()

45877

In [173]:
45877 - 44464 - 725 

688

#### values > 1
* Problem: same article multiple times in same basket

In [ ]:
wrong_retp = orders_features[orders_features['min_5_top_5_sqrt_w_conf']>1]
wrong_retp

In [178]:
# all values in train data
"""over1_index = orders_features[orders_features['min_5_top_5_sqrt_w_conf']>1].index
for index in over1_index:
    print(index)
    order_id = orders.loc[index]["basketID"]
    artnr = orders.loc[index]["itemID"]
    if any(orders[orders["basketID"]==order_id]["itemID"].duplicated()):
        print("index duplicated")
        if any(orders[(orders["basketID"]==order_id)&(orders["itemID"]==artnr)]["itemID"].duplicated()):
            print("art")"""

'over1_index = orders_features[orders_features[\'min_5_top_5_sqrt_w_conf\']>1].index\nfor index in over1_index:\n    print(index)\n    order_id = orders.loc[index]["basketID"]\n    artnr = orders.loc[index]["itemID"]\n    if any(orders[orders["basketID"]==order_id]["itemID"].duplicated()):\n        print("index duplicated")\n        if any(orders[(orders["basketID"]==order_id)&(orders["itemID"]==artnr)]["itemID"].duplicated()):\n            print("art")'

In [396]:
orders_origin = orders.groupby(["basketID", "itemID"]).size().reset_index()
orders_origin[orders_origin[0]>1]
# sum of transactions with 2/ >3 same articles in basket
orders_origin[orders_origin[0]==2][0].sum()
orders_origin[orders_origin[0]>2][0].sum()
51253/orders.shape[0]
426221 - 81591 +135144+51253 # 
(135144+51253)/orders.shape[0]

531027

In [403]:
"""dual_rets = orders_origin[orders_origin[0]==2].apply(lambda x: orders[(orders["basketID"]==x["basketID"])&
                                                         (orders["itemID"]==x["itemID"])][ret_col].sum()/2, axis=1)
dual_rets.mean()"""
# 0.5367903865506423

0.5367903865506423

In [357]:
"""multi_rets = orders_origin[orders_origin[0]>2].apply(lambda x: orders[(orders["basketID"]==x["basketID"])&
                                                         (orders["itemID"]==x["itemID"])][ret_col].sum()/x[0], axis=1)
multi_rets.mean()"""
# 0.48830894072441255

0.48830894072441255

### problem: count multi-arts in basket
* lookup dict creation
* Code: get_ret_perc_train lookup

In [367]:
multi_arts = orders[orders["basketID"].duplicated(keep=False)]

In [368]:
ret_col= parameters['ret_col']
ordnr_col= parameters['ordnr_col']
artnr_col= parameters['artnr_col']

In [379]:
multi_arts = multi_arts[[artnr_col, ordnr_col, ret_col]]

# 1706 transactions of non- retarts
ret_arts = multi_arts[multi_arts[ret_col]==1][artnr_col].unique()

art_df = multi_arts[multi_arts[artnr_col]==29][[ordnr_col, ret_col]] # has duplicates
ret_orders = art_df[art_df[ret_col]>=1].drop_duplicates()[ordnr_col]
basket_df = multi_arts[multi_arts[ordnr_col].isin(art_df[ordnr_col])]
ret_incides = basket_df[basket_df[ordnr_col].isin(ret_orders)].index
basket_df.loc[ret_incides, ret_col] = 1
basket_df = basket_df.drop(ret_orders.index)

basket_grouped = basket_df.groupby(artnr_col).size()
arts_basket_occ = basket_grouped[basket_grouped>=5]
subdf[(subdf["itemID"]==29)&(subdf["returnShipment"]==0)]

#### get_ret_perc_train lookup

In [401]:
bestnr = "26002012-04-04"
ret_ind = 0
data = train[["itemID", "basketID", "returnShipment"]]
ret_p_art = lookup_basket_ret_p[5][29]
occ_arts = lookup_basket_occ[5][29]
sort_arts = lookup_lift_arts[5][29]
top_f = 5
weights = "sqrt_w"
# problem for multi-variants on arts
    # with no return

In [284]:
# art_df[art_df["basketID"].isin(ret_orders)] # has 123 entries
# ret_orders = art_df[art_df[ret_col]>=1][ordnr_col] # has 119 entries
ret_incides = basket_df[basket_df[ordnr_col].isin(ret_orders)].index

In [402]:
basket_arts = data[(data.iloc[:, 1] == bestnr) &    # iloc[:,1] = Column orderID -> same order arts
         (data.iloc[:, 0].isin(ret_p_art.index))].iloc[:, 0]     # i
sorted_arts = sort_arts.loc[basket_arts].sort_values().index.to_series()
sorted_arts

itemID
155    155
213    213
214    214
29      29
29      29
Name: itemID, dtype: int64

In [406]:
# (ret_p_art.loc[sorted_arts] * occ_basket)/(occ_basket-1)
ret_p_basket = ret_p_art.loc[sorted_arts]
occ_basket = occ_arts.loc[sorted_arts]
len(orders_features[orders_features["min_5_top_5_sqrt_w_conf"]>1]["returnShipment"])
 # more nas for lift

In [188]:
len(orders_features[orders_features["min_5_top_5_sqrt_w_lift"]>1]["returnShipment"])

25

In [ ]:
def get_ret_perc_train(bestnr, ret_ind, data, ret_p_art, occ_arts, sort_arts, top_f, weights):
    # sort arts: use df with lift if lift, else no_occurances
    # function init: pick all articles from basket 
    basket_arts = data[(data.iloc[:, 1] == bestnr) &    # iloc[:,1] = Column orderID -> same order arts
             (data.iloc[:, 0].isin(ret_p_art.index))].iloc[:, 0]     # iloc[:,0] = Column article -> has entry in ret_p_art
    if len(basket_arts)==0:
        return np.nan, 0
    else: 
        # sorting specified by series in input (from lookup_lift_arts or lookup_basket_occ)
        sorted_arts = sort_arts.loc[basket_arts].sort_values().index.to_series()

        if top_f:
            if type(top_f) == str: # for values like var_5 -> variable 5 or if len(arts)>5 -> take half of the most frq
                n_min = [int(s) for s in top_f if s.isdigit()][0]
                top_f = max(n_min, round(len(sorted_arts)/2))
            sorted_arts = sorted_arts.iloc[-top_f:]

        # adapt for training data (fix ret_p) to prevent overfitting
        ret_p_basket = ret_p_art.loc[sorted_arts]
        occ_basket = occ_arts.loc[sorted_arts]
        ret_p_basket = (ret_p_basket*occ_basket-ret_ind   # numerator = no returns of article
                                )/(occ_basket-1)  # denominator = current order

        if weights: # if no weigths: only use mean of pcts
            weight = range(1, sorted_arts.shape[0]+1)
            if weights == "sqrt_w":
                weight = np.sqrt(weight)
            ret_pcts = (ret_p_basket*weight).sum()/sum(weight)
        else:
            ret_pcts = ret_p_basket.mean()

        return ret_pcts, occ_basket.sum()

In [119]:
# orders_features[orders_features]

In [15]:
orders_features["min_5_top_5_sqrt_w_conf_Imp"].sum()

117439

In [13]:
# orders_features.columns
max(orders_features["min_5_top_5_sqrt_w_conf"])

1.2

In [41]:
# orders.loc[3505]
# orders.loc[292247]

In [39]:
# (train[~train["itemID"].isin(test["itemID"])]["itemID"].value_counts()>0).sum()
# orders_features[orders_features["min_5_top_5_sqrt_w_conf"]>1]

In [27]:
train["itemID"].nunique()

3002

In [26]:
test["itemID"].nunique()

1283

In [33]:
for i, col in enumerate(cols['cat_catboost']):
    print(i, col)

0 manufacturerID
1 salutation
2 state
3 item_color_cat
4 item_size_cat
5 monthOrdered
6 itemID


## Catboost training (using created features + not using)

In [43]:
orders_features = catalog.load("orders_features")
# orders = catalog.load("orders_full")
parameters = catalog.load("parameters")
train = orders_features[orders_features["val_set"]==0]
test = orders_features[orders_features["val_set"]==1]
negate_retp = True
if negate_retp:
    imp_ind = orders_features[orders_features['min_5_top_5_sqrt_w_lift_Imp']==1].index
    orders_features.loc[imp_ind, "min_5_top_5_sqrt_w_lift"] = -1
    print(orders_features[orders_features["min_5_top_5_sqrt_w_lift"]==-1].shape[0])
for add_ret_p in [True]: # False
    cols = parameters["coldict"]
    cat_cols = cols['cat_pass']+cols['cat_catboost'][:6]
    catboosted = (pd.Series(cols['cat_catboost'][6:]) + "Catboost").tolist()
    cont_cols = cols['cont'] # + catboosted
    if add_ret_p:
        print("added")
        cat_cols +=  ['min_5_top_5_sqrt_w_lift_Imp']
        cont_cols += ['min_5_top_5_sqrt_w_lift']
    outpu_col = cols['output_col']

    X_train = train[cat_cols+cont_cols]
    y_train = train[outpu_col]
    X_test_p = test[test["basketNArts"]>1][cat_cols+cont_cols]
    y_test_p = test[test["basketNArts"]>1][outpu_col]
    X_test = test[cat_cols+cont_cols]
    y_test = test[outpu_col]
    categorical_features_indices = np.array(range(0,len(cat_cols)))

    train_pool = Pool(X_train, y_train, cat_features=categorical_features_indices)
    validate_pool = Pool(X_test, y_test, cat_features=categorical_features_indices)

    params = {
        'iterations': 1000,
        'learning_rate': 0.005,
        'eval_metric': metrics.F1(),
        'random_seed': 42,
        'logging_level': 'Silent',
        'use_best_model': True,
        'depth': 10,
        'allow_writing_files': False,
        'od_type': 'Iter',  # https://catboost.ai/en/docs/features/overfitting-detector-desc#od_wait
        'od_wait': 50,
    }
    model = CatBoostClassifier(**params)
    model.fit(train_pool, eval_set=validate_pool)
    y_pred = model.predict(X_test)
    y_pred_p = model.predict(X_test_p)
    print('F1_test: {:.4}'.format(
        f1_score(y_test, y_pred)),
          'F1_train: {:.4}'.format(
        f1_score(y_train, model.predict(X_train))),
         'MCC_test: {:.4}'.format(
        matthews_corrcoef(y_test, y_pred)),
          'Recall_test: {:.4}'.format(
        recall_score(y_test, y_pred)),
          'Prec_test: {:.4}'.format(
        precision_score(y_test, y_pred)),
          'MCC_test_multiarts: {:.4}'.format(
        matthews_corrcoef(y_test_p, y_pred_p)),
          'F1_test_multiarts: {:.4}'.format(
        f1_score(y_test_p, y_pred_p)),
          'Recall_test_multiarts: {:.4}'.format(
        recall_score(y_test_p, y_pred_p)),
         'Prec_test_multiarts: {:.4}'.format(
        precision_score(y_test_p, y_pred_p)))
    print(X_train.columns)
    imp_dict = {}
    feat_imp = model.get_feature_importance()
    for i, entry in enumerate(X_train.columns.tolist()):
        imp_dict[entry] = feat_imp[i]
    print(imp_dict)
    print("\n")

2022-02-12 15:06:51,171 - kedro.io.data_catalog - INFO - Loading data from `orders_features` (PickleDataSet)...
2022-02-12 15:06:51,209 - kedro.io.data_catalog - INFO - Loading data from `parameters` (MemoryDataSet)...
117439
added
F1_test: 0.6669 F1_train: 0.65 MCC_test: 0.2943 Recall_test: 0.7092 Prec_test: 0.6294 MCC_test_multiarts: 0.2876 F1_test_multiarts: 0.6819 Recall_test_multiarts: 0.7429 Prec_test_multiarts: 0.6301
Index(['weekdayOrdered', 'weekdayReceived', 'weDelivery', 'accOldSys', 'color',
       'size', 'userAgeImp', 'manufacturerID', 'salutation', 'state',
       'item_color_cat', 'item_size_cat', 'monthOrdered',
       'min_5_top_5_sqrt_w_lift_Imp', 'discount', 'basket_price',
       'basketNArts', 'durationDelivery', 'accAgeAtOrder', 'UserAgeOrder',
       'min_5_top_5_sqrt_w_lift'],
      dtype='object')
{'weekdayOrdered': 7.107467985335478, 'weekdayReceived': 17.195172736123652, 'weDelivery': 0.0, 'accOldSys': 0.2989772989769255, 'color': 0.0784513626532197, 'size':

F1_test: 0.6663 F1_train: 0.6547 MCC_test: 0.2763 Recall_test: 0.7276 Prec_test: 0.6145 MCC_test_multiarts: 0.272 F1_test_multiarts: 0.6786 Recall_test_multiarts: 0.7485 Prec_test_multiarts: 0.6207
Index(['weekdayOrdered', 'weekdayReceived', 'weDelivery', 'accOldSys', 'color',
       'size', 'userAgeImp', 'manufacturerID', 'salutation', 'state',
       'discount', 'basket_price', 'basketNArts', 'durationDelivery',
       'accAgeAtOrder', 'UserAgeOrder', 'item_color_catCatboost',
       'item_size_catCatboost', 'monthOrderedCatboost', 'itemIDCatboost'],
      dtype='object')
{'weekdayOrdered': 7.785037473298434, 'weekdayReceived': 20.91562459237241, 'weDelivery': 0.0, 'accOldSys': 0.0, 'color': 0.0, 'size': 0.0, 'userAgeImp': 0.0, 'manufacturerID': 0.0, 'salutation': 0.0, 'state': 0.0, 'discount': 0.7344126285195848, 'basket_price': 20.282720428204538, 'basketNArts': 2.739845492558018, 'durationDelivery': 3.1150894324060823, 'accAgeAtOrder': 0.0, 'UserAgeOrder': 0.0, 'item_color_catCatboost': 4.546012982523614, 'item_size_catCatboost': 39.88125697011731, 'monthOrderedCatboost': 0.0, 'itemIDCatboost': 0.0}


added
F1_test: 0.672 F1_train: 0.6377 MCC_test: 0.2959 Recall_test: 0.7256 Prec_test: 0.6259 MCC_test_multiarts: 0.2911 F1_test_multiarts: 0.6857 Recall_test_multiarts: 0.7534 Prec_test_multiarts: 0.6293
{'weekdayOrdered': 6.0432872136628575, 'weekdayReceived': 14.056111304357628, 'weDelivery': 0.04382915578406688, 'accOldSys': 0.6872855946561911, 'color': 0.10140940999754865, 'size': 0.6295100879362566, 'userAgeImp': 0.0, 'manufacturerID': 1.0800540143285824, 'salutation': 0.5589790628646327, 'state': 0.2928061550647946, 'item_color_cat': 0.7363794010498242, 'item_size_cat': 0.0, 'monthOrdered': 1.8558962767811982, 'min_5_top_5_sqrt_w_lift_Imp': 2.612745147533874, 'discount': 0.7761642017951633, 'basket_price': 16.959253026071767, 'basketNArts': 3.5402049722338953, 'durationDelivery': 7.113492550822126, 'accAgeAtOrder': 0.5860725071122737, 'UserAgeOrder': 1.78714762863198, 'itemIDCatboost': 37.746591238624376, 'min_5_top_5_sqrt_w_lift': 2.792781050690829}